hedera-node/src/main/java/com/hedera/services/ledger/interceptors/StakingAccountsCommitInterceptor.java

```diff
@@ -228,8 +230,16 @@ private void finalizeStakeMetadata(final EntityChangeSet<AccountID, MerkleAccoun
 						roundedToHbar(stakeToAward),
 						finalDeclineRewardGiven(account, changes));
 			}
+			stakeMetaChanged = hasStakeMetaChanges(changes);
+			if (stakeMetaChanged) {
+				stakeAtStartOfLastRewardedPeriodUpdates[i] = -1;
+			} else if (rewardsEarned[i] > 0 ||
```

Comment:
We should extract this complex condition into a method, e.g.
```
else if (shouldRememberStakeStartFor(account, curStakedId, rewardsEarned[i])) {
    stakeAtStartOfLastRewardedPeriodUpdates[i] = account.totalStake();
}

...

private boolean shouldRememberStakeStartFor(
    @NotNull final MerkleAccount account,
    final long curStakedId,
    final long reward
) {
    if (curStakedId >= 0 || account.isDeclinedReward()) {
        // Alice cannot receive a reward for today, so nothing to remember here
        return false;
    }
    if (reward < 0) {
        // Alice's stake could not have changed this transaction, so nothing to remember
        return false;
    } else if (reward > 0) {
        // Alice earned a reward without changing her stake metadata, so she is still eligible
        // for a reward today; since her total stake will change this txn, we remember its 
        // current value to reward her correctly for today no matter what happens later on
        return true;
    } else {
        if (account.getStakeAtStartOfLastRewardedPeriod() != -1) {
            // Alice was in a reward situation, but did not earn anything because she already
            // received a reward earlier today; we preserve our memory of her stake from then
            return false;
        }
        // Alice was in a reward situation for the first time today, but received nothing---
        // either because she is staking < 1 hbar, or because she started staking only 
        // today or yesterday; we don't care about the exact reason, we just remember 
        // her total stake as long as she didn't begin staking today exactly
        return account.getStakePeriodStart() < stakePeriodManager.currentStakePeriod();
    }
}
```